packages/workers-shared/asset-worker/src/handler.ts

```diff
@@ -0,0 +1,589 @@
+		// special handling so you don't drop / if the path is just /
+		if (pathname === "/index.html") {
+			if (
+				(redirectResult = await safeRedirect(
+					"/index.html",
+					"/",
+					configuration,
+					exists,
+					skipRedirects
+				))
+			) {
+				return redirectResult;
+			}
+		} else if (
+			(redirectResult = await safeRedirect(
+				pathname,
+				pathname.slice(0, -"/index.html".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo/index.html exists so redirect to /foo
+			return redirectResult;
+		} else if (exactETag) {
+			// there's both /foo.html and /foo/index.html so we serve /foo/index.html at /foo/index.html only
+			return { asset: { eTag: exactETag, status: 200 }, redirect: null };
+		} else if (
+			(redirectResult = await safeRedirect(
+				`${pathname.slice(0, -"/index.html".length)}.html`,
+				pathname.slice(0, -"/index.html".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo.html exists so redirect to /foo
+			return redirectResult;
+		}
+	} else if (pathname.endsWith("/")) {
+		if (pathname === "/") {
+			if ((eTagResult = await exists("/index.html"))) {
+				// /index.html exists so serve at /
+				return { asset: { eTag: eTagResult, status: 200 }, redirect: null };
+			}
+		} else if (
+			(redirectResult = await safeRedirect(
+				`${pathname.slice(0, -"/".length)}.html`,
+				pathname.slice(0, -"/".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo.html exists so redirect to /foo
+			return redirectResult;
+		} else if (
+			(redirectResult = await safeRedirect(
+				`${pathname.slice(0, -"/".length)}/index.html`,
+				pathname.slice(0, -"/".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo/index.html exists so redirect to /foo
+			return redirectResult;
+		}
+	} else if (pathname.endsWith(".html")) {
+		if (
+			(redirectResult = await safeRedirect(
+				pathname,
+				pathname.slice(0, -".html".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo.html exists so redirect to /foo
+			return redirectResult;
+		} else if (
+			(redirectResult = await safeRedirect(
+				`${pathname.slice(0, -".html".length)}/index.html`,
+				pathname.slice(0, -".html".length),
+				configuration,
+				exists,
+				skipRedirects
+			))
+		) {
+			// /foo/index.html exists so redirect to /foo
+			return redirectResult;
+		}
+	}
+
+	if (exactETag) {
+		// there's a binary /foo file
+		return { asset: { eTag: exactETag, status: 200 }, redirect: null };
+	} else if ((eTagResult = await exists(`${pathname}.html`))) {
+		// /foo.html exists so serve at /foo
+		return { asset: { eTag: eTagResult, status: 200 }, redirect: null };
+	} else if ((eTagResult = await exists(`${pathname}/index.html`))) {
+		// /foo/index.html exists so serve at /foo
+		return { asset: { eTag: eTagResult, status: 200 }, redirect: null };
```

Comment:
Not a blocker for merge, but `{ asset: { eTag: eTagResult, status: 200 }, redirect: null }` is used quite a lot, and could be extracted into a shared variable